### PR TITLE
feat: add benchmarking suite and fix label parsing in dataset features

### PR DESCRIPTION
**Pull request by ddevin96**
### All submissions

- [x] Have you followed the guidelines in our [Contributing](../../CONTRIBUTING.md) document?
- [x] Have you checked to ensure there are no other open [Pull Requests](../../../pulls) for the same changes?
- [x] Have you made sure you have rebased your branch to the latest commit on the target branch?

### Description

Please include a summary of the changes.

### Checklist

- [x] Does your submission pass all tests? (use `make test`)
- [x] Have you written tests to cover all your changes? If not, provide a reason.
- [x] Have you lint your code locally before submission? (use `make format`)
- [x] Have you type checked your code locally before submission? (use `make typecheck`)

### CODE REVIEW

File: benchmark/bench_hyperlink_prediction.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                train_dataset,
+                val_dataset,
+                test_dataset,
+                num_nodes,
+                num_features,
+            )
+
+            print(f"Running benchmark for dataset: {dataset_name}")
+
+            data_loader = DataLoader.from_datasets(
```
Comment (tizianocitro) [inline benchmark/bench_hyperlink_prediction.py:73]: We should pass a None test_dataset here as we should build it separately. We should use sample_full_hypergraph=True in the test set. The data module supports None.

File: benchmark/bench_hlp.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    test_set_negative_ratio = args.test_set_negative_ratio
+    split_ratios = args.split_ratios
+    datasets = args.datasets
+    task = args.task
+
+    print("Loading and preparing datasets...")
+    prepared_datasets = {}
+    for r in range(run):
+        for dataset_name in datasets:
+            picked_seed = seed[r]  # ogni run usa lo stesso seed
```
Comment (tizianocitro) [inline benchmark/bench_hlp.py]: Italian comment.

File: benchmark/bench_hlp.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+                        num_run=r,
+                        train_hyperedge_index=train_dataset.hdata.hyperedge_index,
+                        max_epochs=max_epochs,
+                    )
+                # model = config[0].model
+                print("Starting training and evaluation...")
+
+                with MultiModelTrainer(
+                    model_configs=config,
+                    default_root_dir=f"benchmark/results/{dataset_name}/{config[0].name}",
```
Comment (tizianocitro) [inline benchmark/bench_hlp.py]: Is it fine if we remove config[0].name} from here? I think it might make things a bit confusing. Let me know in case.

File: benchmark/bench_nc.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    datasets = args.datasets
+    task = args.task
+
+    print("Running task:", task)
+
+    print("Loading and preparing datasets...")
+    prepared_datasets = {}
+    for r in range(run):
+        for dataset_name in datasets:
+            picked_seed = seed[r]  # ogni run usa lo stesso seed
```
Comment (tizianocitro) [inline benchmark/bench_nc.py]: Ditto.

File: benchmark/bench_nc.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                num_nodes,
+                num_features,
+            )
+
+            print(f"Running benchmark for dataset: {dataset_name}")
+
+            data_loader = DataLoader.from_datasets(
+                train_dataset=train_dataset,
+                val_dataset=val_dataset,
+                test_dataset=test_dataset,
```
Comment (tizianocitro) [inline benchmark/bench_nc.py]: Same here for the test set.

File: benchmark/bench_nc.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+                "hypergcn_no_mediator",
+                "hypergcn_with_mediator",
+                "mlp",
+                # "nhp",
+                "villain_node",
+                # "villain_hyperedge",
+                "node2vec",
+            ]
+
+            max_epochs = 100
```
Comment (tizianocitro) [inline benchmark/bench_nc.py]: This should be set per model in their model config via trainer_kwargs.

File: benchmark/bench_nc.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                        train_hyperedge_index=train_dataset.hdata.hyperedge_index,
+                        max_epochs=max_epochs,
+                        num_classes=num_classes,
+                    )
+                # model = config[0].model
+                print("Starting training and evaluation...")
+
+                with MultiModelTrainer(
+                    model_configs=config,
+                    default_root_dir=f"benchmark/results/{dataset_name}/{config[0].name}",
```
Comment (tizianocitro) [inline benchmark/bench_nc.py]: Ditto for the model name.

File: benchmark/common_hlp.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+    k_nodes: int = 2,
+    seed: int = 42,
+    test_set_negative_ratio: float = 0.6,
+    task: TaskEnum = TaskEnum.HYPERLINK_PREDICTION,
+) -> tuple[Dataset, Dataset, Dataset, int, int]:
+
+    dataset = get_dataset_by_name(dataset_name=dataset_name, task=task)
+    dataset.remove_hyperedges_with_fewer_than_k_nodes(k=k_nodes)
+
+    num_features = dataset.hdata.x.shape[1] if dataset.hdata.x is not None else 32
```
Comment (tizianocitro) [inline benchmark/common_hlp.py]: dataset.hdata.x is never None, so this might be a bug where we always have 1 as node feature embedding size. The reason is that we default to a x of shape (num_nodes, 1).

File: benchmark/common_nc.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+    task: TaskEnum = TaskEnum.HYPERLINK_PREDICTION,
+) -> tuple[Dataset, Dataset, Dataset, int, int]:
+
+    dataset = get_dataset_by_name(dataset_name=dataset_name, sampling_strategy="node", task=task)
+    dataset.remove_hyperedges_with_fewer_than_k_nodes(k=k_nodes)
+    # dataset.hdata.y = node_labels_from_node_degrees(
+    #     node_incidences=dataset.hdata.hyperedge_index[0],
+    #     num_nodes=dataset.hdata.num_nodes,
+    # )
+    num_features = dataset.hdata.x.shape[1] if dataset.hdata.x is not None else 32
```
Comment (tizianocitro) [inline benchmark/common_nc.py]: Ditto.

File: hypertorch/data/supported_datasets.py [hunk 10]
```diff
@@ -111,12 +111,17 @@ def list_datasets() -> list[str]:
...
+    dataset_name: str,
+    task: Task = TaskEnum.HYPERLINK_PREDICTION,
+    sampling_strategy: SamplingStrategy = SamplingStrategyEnum.HYPEREDGE,
+) -> Dataset:
     """
     Instantiate a supported dataset by name.
 
     Args:
         dataset_name: Name returned by ``list_datasets``.
+        task: Learning task used to initialize the dataset.
```
Comment (tizianocitro) [inline hypertorch/data/supported_datasets.py:124]: Missing strategy arg.

**Review by tizianocitro — CHANGES_REQUESTED** [review]
I think there are two important things to look at:
- The test set configuration that does not have sample_full_hypergraph.
- The num_feature configuration in the prepare methods.

**Comment by tizianocitro** [comment]
I noticed that we need to assign a separate name to the fast version of HyperGCN, otherwise results will overwrite the normal ones.

**Review by tizianocitro — APPROVED** [review]
